ddtrace/internal/datadog/profiling/dd_wrapper/include/profile.hpp

```diff
@@ -0,0 +1,82 @@
+#pragma once
+
+#include "constants.hpp"
+#include "types.hpp"
+
+#include <atomic>
+#include <deque>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <string_view>
+#include <unordered_set>
+#include <vector>
+
+extern "C"
+{
+#include "datadog/profiling.h"
+}
+
+namespace Datadog {
+
+// Unordered containers don't get heterogeneous lookup until gcc-10, so for now use this
+// strategy to dedup + store strings.
+using StringTable = std::unordered_set<std::string_view>;
+
+// Serves to collect individual samples, as well as lengthen the scope of string data
+class Profile
+{
+  private:
+    // Serialization for static state
+    // - string table
+    // - ddog_profile
+    std::atomic<bool> first_time{ true };
+    std::mutex profile_mtx{};
+    std::atomic<pid_t> current_pid{ 0 };
+
+    // Storage for strings
+    std::deque<std::string> string_storage{};
+    StringTable strings{};
+    std::mutex string_table_mtx{};
+
+    // Configuration
+    SampleType type_mask{ 0 };
+    unsigned int max_nframes{ g_default_max_nframes };
+    ddog_prof_Period default_period{};
+
+    // Sampler setup
+    void setup_samplers();
+
+    // Lookup for values
+    ValueIndex val_idx{};
+
+    // Configuration for the pprof exporter
+    std::vector<ddog_prof_ValueType> samplers{};
+
+    // These are initialized here as skeleton objects, but they cannot be used until
+    // they're initialized by libdatadog
+    ddog_prof_Profile cur_profile{};
+    ddog_prof_Profile last_profile{};
```

Comment:
Why not init in the constructor?

Reply:
This should get tightened up.  Will add some hooks to this in tests.